interface/MO_Interface.h

```diff
@@ -0,0 +1,173 @@
+   * implementing this interface must ensure that it will call the DMClient::XXX() method 
+   * exactly once for each time this callback is invoked by the client library and the
+   * callback returned true.
+   *
+   * For each invocation of the execute callback by the client library the implementation
+   * can call DMClient::XXX() either from inside its implementation of the callback
+   * for synchronous reporting, or at *any* time after returning from the callback 
+   * (see section 5.4.1 of the OMA DM 2.0 protocol specification) 
+   * 
+   * TODO: check if additional info (parameters) are needed in order
+   * to correctly link the response (synchronous or asynchronous) with the command.
+   * 
+   * This interface is not considered stable yet
+   */
+  virtual bool execute(const std::string node_path) = 0;
+
+
+  /**
+   *  @}
+   *  @{
+   *  (2) callbacks called by the protocol client library as part of Session or MO management
+   */
+
+  /**
+   * @brief callback for the protocol client library to check MO compatibility
+   *
+   * The protocol client library will call this before adding an instance of an MO implementation
+   * to its MO tree in order to check if the given implementation of the MO class is compatible with
+   * the DDF file registered with the client library. 
+   *
+   * @param[in] ddfname - DDFName as given in the respective name in a DDF file. Example: "urn:oma:mo:oma-dm-dmacc:1.2"
+   * @return - MO implementation classes should only return true if they are compatible with the DDF file corresponding to the given DDFName
+   * 
+   */
+  virtual bool check_ddf_name_compatibility(std::string ddfname) = 0;
+
+  /**
+   * @brief callback on adding the MO instance to the client library's MO tree
+   *
+   * The protocol client library will call this after successfully adding the instance of the MO
+   * into its MO tree.
+   */
+  virtual void init_mo() = 0; // gets called by the client after adding an instance to the tree.
+
+  /**
+   * @brief callback on removing the MO instance from the client library's MO tree
+   *
+   *  Use currently unclear. May be removed later if not needed
+   */
+  virtual void close_mo() = 0; // unclear
+
+  /**
+   * @brief callback to read the miid of the MO instance
+   *
+   * The protocol client library will call this to retrieve the miid of a given MO instance.
+   * The MO classes implementing this interface are responsible for ensuring that miid's are 
+   * unique between all instances of this MO type.
+   */
+  virtual std::string miid() const = 0;
+
+  /**
+   * @provide json object for MO content serialization
+
+   * This shall return a JSON object suitable for use in serialization according to the format
+   * described in OMA-DM Protocol Spec V2.0 section 7.2.1.4 - management object serialization.
+   * The output of this method shall correspond to the "MOData" object in that specification.
+   */
+  virtual nlohmann::json serialize_json() const = 0;
```

Comment:
same as the "miid", it is the responsibility of client lib to serialize the MO tree into a json format.
my understanding is that these interfaces are defined for client app, which implement them. So I can't get the point of last two ones.

Reply:
SDK needs the json representation of the data to generate some protocol packages. The SDK could also accept some other data format of the MO data (let's say, a struct of "Node" datatypes like used in the MO_CachedBase class) and build the json in the sdk. But some data type needs to be agreed on. And since the interface of nlohmann::json is very convenient, I thought it actually easier to handle for the MO implementation instead of building a tree of struct...

Reply:
I've thought about this some more, and I generally agree that injecting the dependency of json lib into the local application side is not really clean. I would like to remove that if I am convinced I have a better solution.

The issue is: While I dislike enforcing json use on the local application as an exchange format with the SDK, I think it is even worse to enforce a predefined data structure that the local application (and their MO implementation) should use internally. So, the "Node" class in the MO_BaseCached Base class that user applications *can* use as a base for their MO implementation is meant as an *offer* only, not as an enforcement. If anyone prefers to use a different datastructure, or even decides not to hold node values in memory at all (because they represent, for example, volatile signal values or continuously growing log or trace files etc...) they should be free to do so.

If I remove this json dependency (and don't replace it with another arbitrary data structure for the same use case), then I need  the client SDK to iterate over the entire tree using separate get callbacks to retrieve MO tree content whenever I need it. That could cause hundreds of calls. Do you think that would be the better solution? I am currently really undecided (for me it is the choice between two bad options), so your opinion would probably make my decision. Or do you have a completely different idea?